### ISSUE TITLE
🚀 [Feature]: Ability to change param values

### ISSUE THREAD

**Issue by MarcusSanchez** [gofiber/fiber#2487]
### Feature Description

When using c.Params() sometimes a function is necessary to format the param value to be further used, which is why the ability to be able to edit the c.Params() value would be a killer feature. Below is a middleware example of how this can be implemented.

### Code Snippet (optional)

```go
func replaceHyphens(c *fiber.Ctx) error {
	if err := c.Next(); err != nil {
		fmt.Println("error: ", err)
                return err
	}
	for _, param := range c.Route().Params {
		value := c.Params(param)
		value = strings.ReplaceAll(value, "-", " ")
                // Something like this that will set the param value before the handler uses it...
		c.SetParam(param, value) // SetParam() doesn't exist
	}
	return nil
}
```

### Checklist:

- [X] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/master/.github/CODE_OF_CONDUCT.md).
- [X] I have checked for existing issues that describe my suggestion prior to opening this one.
- [X] I understand that improperly formatted feature requests may be closed without explanation.

**Comment by luk3skyw4lker** [comment]
I don't think that this should be a feature since it wouldn't remove the need for a function to format the parameters, it would only move where it is placed, like instead of placing it in the fetching of the params it would be in a middleware and in both cases the responsability of the params formatting is of the service using the framework, not the framework itself.

**Comment by MarcusSanchez** [comment]
> I don't think that this should be a feature since it wouldn't remove the need for a function to format the parameters, it would only move where it is placed, like instead of placing it in the fetching of the params it would be in a middleware and in both cases the responsability of the params formatting is of the service using the framework, not the framework itself.
> 
> But in the case this really becomes a valid feature proposal I would like to implement it.

Well of course there will always be a need for a function to format params in the need to query a database or take care of differing type cases, however the need to call it in every handler would disappear as the middleware will self invoke. Which means essentially instead of having to value := format(c.Params("param") every function, c.Params("param") would already be formatted. This is would be good for me as in my rest apis I always must call it like this:

err := helpers.DeleteTeam__(strings.ReplaceAll(c.Params("teamName"), "-", " "))

Now, based on my understanding of Express and most likely Fiber, getting access to a handler's params/request would require the calling of c.Next() which means the handler would run before you are able to make any changes to params which is why this feature request would be hard to implement. It would require creating a way to gain access to the request body/ params without calling c.Next().

**Comment by ReneWerner87** [comment]
I think I had misread the feature in the previous article
https://github.com/gofiber/fiber/issues/2480#issuecomment-1567850450

is there such a feature already in other frameworks? (can you give examples)
looks pretty custom to me, as if it could be a general feature

normally i would create a helper in the project where i want to replace these characters and use it in the functions as a wrapper for the function or make a change in the transfer directly to the consumer of the api

**Comment by luk3skyw4lker** [comment]
Giving the Express example you can do this and directly alter the req.params object. I have already done this using a middleware, the only thing you should pay attention is that the middleware to do this should be defined before the final request handler. The parsing of URL params/query params is done way before the final handler.

I think it wouldn't be that hard to implement it, since the function would "only" alter the parameters value. If Fiber is based on Express at that level, I don't think it would be that big of a deal.

**Comment by MarcusSanchez** [comment]
> I think I had misread the feature in the previous article [#2480 (comment)](https://github.com/gofiber/fiber/issues/2480#issuecomment-1567850450)
> 
> is there such a feature already in other frameworks? (can you give examples) looks pretty custom to me, as if it could be a general feature
> 
> normally i would create a helper in the project where i want to replace these characters and use it in the functions as a wrapper for the function or make a change in the transfer directly to the consumer of the api

Well being Fiber is my first backend framework that I've picked up, I'm not to familiar with other examples outside of Fiber/Express, however I have done this in Flask with Python, which is where I've gotten the idea to do it here. It looks something like this:

@app.before_request
def replace_hyphens():
    value = request.view_args.get("value")
    value = value.replace("-", " ")
    request.view_args["value"] = value


@app.route("/<value>")
def serve_static(value):
    return value
    

This works as I would like it to in Fiber, but remember this isn't a crazy big issue, in fact I could very well live without it. Though it would be nice.

**Comment by luk3skyw4lker** [comment]
@ReneWerner87 you think this should go through? If so I will work on it and open a PR later

**Comment by ReneWerner87** [comment]
can you share your idea with us, what would the solution look like ?

**Comment by MarcusSanchez** [comment]
Overall, the abstracted view of a solution would do these things:

1. Obtain a way to gain access to the handler's params, via c.Params() or maybe a new function. This would be great if it also works for query params.
- Possible limitations:
        - Middleware params aren't empty, and can possibly clash with handler params.
        - Gaining access to a handler's params typically require the calling of c.Next(), which would entail that the handler would run before the middleware, which obviously wouldn't work therefore a workaround to gain access would be required.

2. Obtain a way to set or update/overwrite c.Params()'s values which will reflect in the actual handler.

Really, these are the only things needed for this feature request. With the ability to view and update params before a handler is run, essentially what you are left with is the ability to format/authenticate/validate (or whatever) a handler's param values via a custom middleware. Again this is an abstracted view therefore there are definitely more limitations under the hood.

**Comment by luk3skyw4lker** [comment]
@ReneWerner87 I think the solution proposed by @MarcusSanchez is pretty straightforward, if you want some more code details, I would ask you to give me sometime to go through the framework code and show you a more detailed proposal

**Comment by ReneWerner87** [comment]
currently each middleware has its own route and the different handlers can only access the parameters in that route

i am curious how this will be solved without affecting the performance, because the routing is built very lazy
the parameters are only determined from the route when they are needed and can also have different names
```go
app.Use(myOtherHandler())
app.Use("*/:id", myIdHandler())

app.Get("/v1/userManager/:userId", userHandlerV1())
app.Get("/v2/users/:uniqueUser", userHandlerV2())
``` 

you are welcome to create a pull request and we will see if we accept the outlined functionality, currently it is too theoretical to say for sure if we would allow it

**Comment by luk3skyw4lker** [comment]
@ReneWerner87 I would like to join the discord server to send some doubts but the link in the README is broken :/

**Comment by ReneWerner87** [comment]
This link https://gofiber.io/discord

**Comment by luk3skyw4lker** [comment]
I think that this would implicate into a major refactor of the parameters parsing, at least for a global middleware use. It would be possible with minor alterations but only using the middleware in a route dedicated manner, like this:

`
app.Get("/route/:param", formatParam, handler)
app.Get("/route2/:param", formatParam, handler)
`

The refactor at the parameter handling could be a big problem, since it's definitely going to have an impact at the performance of the framework, since the parameters parsing would have to be done globally when the request is received by the handler (I don't see any way to do this besides parsing the parameters globally), and not lazy as it's done today.
